### PR TITLE
Use dump_course_ids_with_filter to limit courses by end date

### PR DESCRIPTION
**Pull request by AbdouSeck**
This should have the exporter start using the `dump_course_ids_with_filter` management command added by [this platform PR](https://github.com/openedx/edx-platform/pull/30070) to limit the number of courses by their end dates. This is controlled by a `time_constraint` configuration value, which should be an integer used to calculate the lower bound for an end date.

### CODE REVIEW

File: exporter/main.py [hunk 1]
```diff
@@ -303,15 +303,40 @@ def match(course):
...
+            'The default value of 3 will be used instead: {e}'
+        )
+        log.warning(warn_msg.format(c=constraint, e=e))
+        constraint = 3
+
+    # Calculate the end date by changing today's year to 3 years ago.
+    # Note that using str on a date object returns a date string with the
+    # format %Y-%m-%d
+    today = datetime.date.today()
+    kwargs['end'] = str(today.replace(year=today.year - constraint))
```
Comment (pwnage101) [inline exporter/main.py]: are you sure this technique always works?  What about leap year dates?  should we be using datetime math instead?
```suggestion
    kwargs['end'] = str(today - datetime.timedelta(days=(365 * constraint)))
```

File: exporter/main.py [hunk 2]
```diff
@@ -303,15 +303,40 @@ def match(course):
...
     kwargs['limit'] = False  # don't limit number of courses
 
+    # Extract the time constraint option and calculate the end date that
+    # should be used to limit the list of returned course IDs.
+    # Please note that the default time constraint is 3 years
+    constraint = str(kwargs.get('time_constraint', '3'))
+    try:
+        constraint = int(constraint)
+    except ValueError as e:
+        # Not sure that this is great idea, but for now, we report the error
```
Comment (pwnage101) [inline exporter/main.py]: why not just halt if there's a value error?

File: exporter/main.py [hunk 3]
```diff
@@ -303,15 +303,40 @@ def match(course):
...
 def get_all_courses(**kwargs):
     log.info('Retrieving all courses')
-
+    # These are options that don't start with the keyword django
+    other_options = ('lms_config', 'studio_config', 'time_constraint')
     # make a set of fixed arguments, so we can memoize
     kwargs = {
         k: v for k, v in kwargs.iteritems()
-        if k.startswith('django') or k == 'lms_config' or k == 'studio_config'
+        if k.startswith('django') or any(k == opt for opt in other_options)
```
Comment (pwnage101) [inline exporter/main.py]: Nit:
```suggestion
        if k.startswith('django') or  k in other_options
```

File: exporter/properties.py [hunk 4]
```diff
@@ -27,7 +27,7 @@
   --orgs=<orgs>            Space separated list of organization identifiers.
                            Can use wildcards.
 """
-
+from __future__ import print_function
```
Comment (pwnage101) [inline exporter/properties.py:30]: :+1:

File: exporter/main.py (hunk 1 above)
Comment (AbdouSeck) [reply exporter/main.py]: I didn't want to deal with figuring whether to use `365` or `366`. Replacing the `year` component with `year - constraint` seemed to make more sense. But I am happy to use `str(today - datetime.timedelta(days=(365 * constraint)))`.

File: exporter/main.py (hunk 3 above)
Comment (AbdouSeck) [reply exporter/main.py]: Not sure what I was thinking. Thanks!

File: exporter/main.py (hunk 2 above)
Comment (AbdouSeck) [reply exporter/main.py]: I added an `ExporterFatalError(Exception)` class and used it to raise an exception when the config value can't be parsed.

File: exporter/main.py (hunk 1 above)
Comment (pwnage101) [reply exporter/main.py]: right, but we're just going for approximately 3 years, and at least we know for sure that it works for every possible date so I still stand by my proposal

File: exporter/main.py [hunk 5]
```diff
@@ -301,17 +309,43 @@ def match(course):
...
+    # Please note that the default time constraint is 3 years
+    constraint = str(kwargs.get('time_constraint', '3'))
+    try:
+        constraint = int(constraint)
+    except ValueError as e:
+        # If the given configuration value cannot be parsed into an integer,
+        # then we quit.
+        error_msg = (
+            'The given time constraint value {c} is not a valid integer. '
+            'The default value of 3 will be used instead: {e}'
```
Comment (pwnage101) [inline exporter/main.py]: remove this obsolete line, default is not used in the case of a fatal exception

File: exporter/main.py [hunk 6]
```diff
@@ -301,17 +309,43 @@ def match(course):
...
+        )
+        raise ExporterFatalError(error_msg.format(c=constraint, e=e))
```
Comment (pwnage101) [inline exporter/main.py]: Nit: use python3.6+ style f-strings, like
```suggestion
        raise ExporterFatalError(f'The given time constraint value {constraint} is not a valid integer.')
```

File: exporter/main.py [hunk 7]
```diff
@@ -301,17 +309,43 @@ def match(course):
...
+
+    # Calculate the end date by changing today's year to 3 years ago.
+    # Note that using str on a date object returns a date string with the
+    # format %Y-%m-%d
+    today = datetime.date.today()
+    kwargs['end'] = str(today - datetime.timedelta(days=(365 * constraint)))
+    if kwargs['end']:
+        msg = 'Limiting the courses to end dates of {d} and above.'
```
Comment (pwnage101) [inline exporter/main.py]: Nit: use "and later" instead of "and above"

File: exporter/main.py (hunk 6 above)
Comment (pwnage101) [reply exporter/main.py]: for simplicity, it's completely acceptable to just raise the stock ValueError class, no need to add complexity by creating a new exception.  This isn't a library or part of a huge codebase, so it's better to keep it simple.
```suggestion
        raise ValueError(f'The given time constraint value {constraint} is not a valid integer.')
```

Comment (AbdouSeck) [reply exporter/main.py]: I used `ValueError`, but since this is python2 codebase, `f-strings` are not supported.

File: exporter/main.py (hunk 7 above)
Comment (AbdouSeck) [reply exporter/main.py]: Updated. Thanks!

**Review by pwnage101 — APPROVED** [review]

**Pull request merged**